### PR TITLE
More datatypes for groupUniqArray() function [#METR-22071]

### PR DESCRIPTION
**Pull request by ludv1x**
Also added additional Arena\* argument for IAggregateFunction::add() function (it allows to make more sophisticated implementations).

### CODE REVIEW

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 1]
```diff
@@ -70,7 +71,7 @@ class AggregateFunctionGroupUniqArray
 		const typename State::Set & set = this->data(place).value;
 		size_t size = set.size();
 		writeVarUInt(size, buf);
-		for (typename State::Set::const_iterator it = set.begin(); it != set.end(); ++it)
+		for (auto it = set.begin(); it != set.end(); ++it)
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Let go further and write `for (const auto & elem : set)`

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 2]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+		for (auto it = set.begin(); it != set.end(); ++it, ++i)
 			data_to[old_size + i] = *it;
 	}
 };
 
+
+/// Generic implementation
+struct AggreagteFunctionGroupUniqArrayGenericData
+{
+	using Set = HashSetWithSavedHash<StringRef, StringRefHash, HashTableGrower<4>, HashTableAllocatorWithStackMemory<16>>;
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Maybe there is inconsistency, because in HashTableGrower, parameter is power of two for space of hash table in number of elements, but in HashTableAllocatorWithStackMemory, parameter is number of bytes.
Please compare with other uses.

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 3]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+	Set value;
+};
+
+template <bool is_plain_column=false>
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Need whitespaces.

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 4]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+class AggreagteFunctionGroupUniqArrayGeneric : public IUnaryAggregateFunction<AggreagteFunctionGroupUniqArrayGenericData, AggreagteFunctionGroupUniqArrayGeneric<is_plain_column>>
+{
+	mutable DataTypePtr input_data_type;
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Why mutable?

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 5]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+	{
+		return std::make_shared<DataTypeArray>(input_data_type->clone());
+	}
+
+	void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+	{
+		auto & set = this->data(place).value;
+		writeVarUInt(set.size(), buf);
+
+		for (auto & elem: set)
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: In all other places, we put whitespace in both sides of colon.

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 6]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+		State::Set & set = this->data(place).value;
+		size_t size;
+		readVarUInt(size, buf);
+		//TODO: set.reserve(size);
+
+		std::string str_buf;
+		for (size_t i = 0; i < size; i++)
+		{
+			readStringBinary(str_buf, buf);
+			set.insert(StringRef(str_buf));
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Dangling StringRef. (Data for strings is not owned.)
Need to test for this method with distributed query processing.

It's worth to add `readStringBinaryInto` method, that reads directly into Arena, by analogy with `readStringInto` method.

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 7]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+
+		bool inserted;
+		State::Set::iterator it;
+
+		StringRef str_serialized = getSerialization(column, row_num, *arena);
+		set.emplace(str_serialized, it, inserted);
+
+		if (!is_plain_column)
+		{
+			if (!likely(inserted))
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Non obvious, that is is likely.
I think, many repeated elements is frequent case too.
Better to remove `likely`.

In fact, benefit from `likely` is very rare and small.

File: dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h [hunk 8]
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
...
+			if (likely(inserted))
+				it->data = arena->insert(str_serialized.data, str_serialized.size);
+		}
+	}
+
+	void merge(AggregateDataPtr place, ConstAggregateDataPtr rhs) const override
+	{
+		auto & cur_set = this->data(place).value;
+		auto & rhs_set = this->data(rhs).value;
+		cur_set.merge(rhs_set);
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h]: Need to check very carefully, that Arena for rhs remains alive.
I don't remember, whether it is true.

File: dbms/include/DB/AggregateFunctions/IAggregateFunction.h [hunk 9]
```diff
@@ -77,7 +78,7 @@ class IAggregateFunction
 	virtual size_t alignOfData() const = 0;
 
 	/// Добавить значение. columns - столбцы, содержащие аргументы, row_num - номер строки в столбцах.
-	virtual void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num) const = 0;
+	virtual void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const = 0;
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/IAggregateFunction.h]: Comment was not edited to document `arena` parameter.

File: dbms/include/DB/AggregateFunctions/IAggregateFunction.h [hunk 10]
```diff
@@ -144,6 +145,11 @@ class IAggregateFunctionHelper : public IAggregateFunction
 	{
 		return __alignof__(Data);
 	}
+
```
Comment (alexey-milovidov) [inline dbms/include/DB/AggregateFunctions/IAggregateFunction.h]: Was not removed.

File: dbms/include/DB/Common/HashTable/HashTable.h [hunk 11]
```diff
@@ -70,6 +70,10 @@ bool check(const T x) { return x == 0; }
 template <typename T>
 void set(T & x) { x = 0; }
 
+bool check(const std::string & x);
```
Comment (alexey-milovidov) [inline dbms/include/DB/Common/HashTable/HashTable.h]: Now useless?

File: dbms/include/DB/Interpreters/Aggregator.h [hunk 12]
```diff
@@ -44,7 +44,7 @@ namespace ErrorCodes
   * Большинство структур данных существует в двух вариантах: обычном и двухуровневом (TwoLevel).
   * Двухуровневая хэш-таблица работает чуть медленнее при маленьком количестве различных ключей,
   *  но при большом количестве различных ключей лучше масштабируется, так как позволяет
-  *  распараллелить некоторые операции (слияние, пост-обработку) естественным образом.
+  * распараллелить некоторые операции (слияние, пост-обработку) естественным образом.
```
Comment (alexey-milovidov) [inline dbms/include/DB/Interpreters/Aggregator.h]: Accidential modification.

File: dbms/include/DB/Interpreters/Aggregator.h [hunk 13]
```diff
@@ -384,7 +384,7 @@ struct AggregationMethodConcat
 		{
 			/** Исправление, если все ключи - пустые массивы. Для них в хэш-таблицу записывается StringRef нулевой длины, но с ненулевым указателем.
 			  * Но при вставке в хэш-таблицу, такой StringRef оказывается равен другому ключу нулевой длины,
-			  *  у которого указатель на данные может быть любым мусором и использовать его нельзя.
+			  * у которого указатель на данные может быть любым мусором и использовать его нельзя.
```
Comment (alexey-milovidov) [inline dbms/include/DB/Interpreters/Aggregator.h]: Same as above.

File: dbms/include/DB/Interpreters/Aggregator.h [hunk 14]
```diff
@@ -977,7 +977,7 @@ class Aggregator
 	void initialize(const Block & block);
 
 	/** Установить блок - пример результата,
-	  *  только если он ещё не был установлен.
+	  * только если он ещё не был установлен.
```
Comment (alexey-milovidov) [inline dbms/include/DB/Interpreters/Aggregator.h]: .

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp [hunk 15]
```diff
@@ -1,13 +1,36 @@
 #include <DB/AggregateFunctions/AggregateFunctionFactory.h>
 #include <DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h>
 #include <DB/AggregateFunctions/Helpers.h>
+#include <common/logger_useful.h>
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp]: May remove.

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp [hunk 16]
```diff
@@ -1,13 +1,36 @@
...
 {
 
 namespace
 {
 
+static IAggregateFunction * createWithExtraTypes(const IDataType & argument_type)
+{
+		 if (typeid_cast<const DataTypeDateTime *>(&argument_type))	return new AggregateFunctionGroupUniqArray<UInt32>;
+	else if (typeid_cast<const DataTypeDate *>(&argument_type)) 	return new AggregateFunctionGroupUniqArray<UInt16>;
+	else
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp]: More clear is to use `DataTypeDateTime::FieldType` and `DataTypeDate::FieldType`.

File: dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp [hunk 17]
```diff
@@ -1,13 +1,36 @@
...
+	{
+		/// Check that we can use plain version of AggreagteFunctionGroupUniqArrayGeneric
+		if (typeid_cast<const DataTypeString*>(&argument_type))
+			return new AggreagteFunctionGroupUniqArrayGeneric<true>;
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp]: And FixedString?

File: dbms/src/Client/Benchmark.cpp [hunk 18]
```diff
@@ -58,11 +59,11 @@ class Benchmark
 	Benchmark(unsigned concurrency_, double delay_,
 			const String & host_, UInt16 port_, const String & default_database_,
 			const String & user_, const String & password_, const String & stage,
-			bool randomize_,
+			bool randomize_, size_t num_repetions_, double max_time_,
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Typo.

File: dbms/src/Client/Benchmark.cpp [hunk 19]
```diff
@@ -206,6 +209,18 @@ class Benchmark
 				report(info_per_interval);
 				watch.restart();
 			}
+
+			if (num_repetions > 0 && info_total.queries >= num_repetions)
+			{
+				std::cout << "The execution is broken since request number of loops is reached\n";
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: In fact, iterations, not loops.
You may pass many queries to benchmark and it will loop many times through them.

File: dbms/src/Client/Benchmark.cpp [hunk 20]
```diff
@@ -183,7 +186,7 @@ class Benchmark
 		Stopwatch watch;
 
 		/// В цикле, кладём все запросы в очередь.
-		for (size_t i = 0; !interrupt_listener.check(); ++i)
+		for (size_t i = 0; !interrupt_listener.check() && (!(num_repetions > 0) || i < num_repetions); ++i)
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: According to next three lines, you cannot specify `num_repetitions` greater than `queries.size()`.

File: dbms/src/Client/Benchmark.cpp [hunk 21]
```diff
@@ -206,6 +209,18 @@ class Benchmark
...
+
+			if (num_repetions > 0 && info_total.queries >= num_repetions)
+			{
+				std::cout << "The execution is broken since request number of loops is reached\n";
+				break;
+			}
+
+			if (max_time > 0 && info_total.watch.elapsedSeconds() >= max_time)
+			{
+				std::cout << "The execution is broken since requested time limit is reached\n";
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: broken means "сломано".

File: dbms/src/Client/Benchmark.cpp [hunk 22]
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
 		desc.add_options()
 			("help", "produce help message")
 			("concurrency,c", boost::program_options::value<unsigned>()->default_value(1), "number of parallel queries")
-			("delay,d", boost::program_options::value<double>()->default_value(1), "delay between reports in seconds")
+			("delay,d", boost::program_options::value<double>()->default_value(1), "delay between reports in seconds (set 0 to disable)")
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Unclear. To disable reports or to disable delay?

File: dbms/src/Client/Benchmark.cpp [hunk 23]
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
...
 			("host,h", boost::program_options::value<std::string>()->default_value("localhost"), "")
 			("port", boost::program_options::value<UInt16>()->default_value(9000), "")
 			("user", boost::program_options::value<std::string>()->default_value("default"), "")
 			("password", boost::program_options::value<std::string>()->default_value(""), "")
 			("database", boost::program_options::value<std::string>()->default_value("default"), "")
 			("stage", boost::program_options::value<std::string>()->default_value("complete"), "request query processing up to specified stage")
+			("loops,l", boost::program_options::value<size_t>()->default_value(0), "number of tests repetions")
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Don't forget about typo.

File: dbms/src/Client/Benchmark.cpp [hunk 24]
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
...
+			("timelimit,t", boost::program_options::value<double>()->default_value(0.), "stop repeating after specified time limit")
```
Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: If many queries, it means not 'repeating' but 'running'.

File: dbms/src/Interpreters/Aggregator.cpp [hunk 25]
```diff
@@ -431,7 +432,7 @@ AggregatedDataVariants::Type Aggregator::chooseAggregationMethod(const ConstColu
 }
 
 
-void Aggregator::createAggregateStates(AggregateDataPtr & aggregate_data) const
+void Aggregator::createAggregateStates(AggregateDataPtr & aggregate_data, Arena * arena) const
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Aggregator.cpp]: No need for `arena` in this method.

File: dbms/src/Interpreters/Aggregator.cpp [hunk 26]
```diff
@@ -1003,6 +1005,7 @@ void NO_INLINE Aggregator::convertToBlockImplFinal(
 	ColumnPlainPtrs & final_aggregate_columns,
 	const Sizes & key_sizes) const
 {
+	//LOG_DEBUG(log, "convertToBlockImplFinal start");
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Aggregator.cpp]: Was not removed.

File: dbms/src/Interpreters/Aggregator.cpp [hunk 27]
```diff
@@ -1808,6 +1830,8 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
 			aggregate_functions[j]->merge(
 				value + offsets_of_aggregate_states[j],
 				(*aggregate_columns[j])[i]);
+
+		LOG_DEBUG(log, "mergeStreamsImplCase");
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Aggregator.cpp]: .

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/src/Client/Benchmark.cpp (hunk 24 above)
Comment (ludv1x) [inline dbms/src/Client/Benchmark.cpp]: I deleted Benchmark.cpp changes from this PR.

There are some additional problems with current version of the benchmark.
I will resolve them and make separated PR.

Comment (alexey-milovidov) [inline dbms/src/Client/Benchmark.cpp]: Ok.

File: dbms/include/DB/Functions/FunctionsMiscellaneous.h [hunk 28]
```diff
@@ -1245,7 +1245,7 @@ class FunctionRunningAccumulate : public IFunction
 		const auto & states = column_with_states->getData();
 		for (const auto & state_to_add : states)
 		{
-			agg_func.merge(place.get(), state_to_add);
+			agg_func.merge(place.get(), state_to_add, nullptr); /// Empty arena!
```
Comment (alexey-milovidov) [inline dbms/include/DB/Functions/FunctionsMiscellaneous.h]: Need to create Arenas. You may use ObjectPool (our ObjectPool, not from Poco) for this task. See usages.

Comment (alexey-milovidov) [inline dbms/include/DB/Functions/FunctionsMiscellaneous.h]: Or add 'bool allocatesMemoryInArena()' method to IAggregateFunction and throw exception for that functions here.

File: dbms/src/Interpreters/Aggregator.cpp [hunk 29]
```diff
@@ -1476,9 +1488,13 @@ void NO_INLINE Aggregator::mergeBucketImpl(
 	{
 		AggregatedDataVariants & current = *data[i];
 
+		/// Select Arena to avoid race conditions
+		Arena * arena = res->aggregates_pools.at(static_cast<size_t>(bucket) % size).get();
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Aggregator.cpp]: Need to create more aggregates_pools if has less than max_threads of them.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql [hunk 30]
```diff
@@ -0,0 +1,10 @@
+DROP TABLE IF EXISTS test.group_uniq;
+CREATE TABLE test.group_uniq ENGINE = Memory AS SELECT number % 10 as id, toString(intDiv((number%10000), 10)) as v FROM system.numbers LIMIT 10000000;
+
+INSERT INTO test.group_uniq SELECT 2 as id, toString(number % 100) as v FROM system.numbers LIMIT 1000000;
+INSERT INTO test.group_uniq SELECT 5 as id, toString(number % 100) as v FROM system.numbers LIMIT 10000000;
+
+SELECT length(groupUniqArray(v)) FROM test.group_uniq GROUP BY id ORDER BY id;
+--SELECT length(groupUniqArray(v)) FROM remote('127.0.0.{1,2,3,4}', 'test', 'group_uniq') GROUP BY id ORDER BY id;
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql]: Why test for distributed is commented?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Comment by alexey-milovidov** [comment]
Need to add test for array of numbers and array of strings types of arguments.

File: dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql (hunk 30 above)
Comment (ludv1x) [inline dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql]: It isn't obvious that two sequential SELECT calls produce output only for first SELECT.
I was lazy to add an two similar test.
Now I use UNION ALL.

**Comment by ludv1x** [comment]
Ok

File: dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql (hunk 30 above)
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql]: It is not true. Look at most of other .sql tests.
All sql tests are run in `multiquery` mode.

Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql]: But when you running query with UNION ALL, you get non-deterministic order of result, when max_threads != 1. So, test will be flappy. Please refine.
You may send another pull request.
